cylc/flow/cycling/iso8601.py

```diff
@@ -765,21 +763,28 @@ def prev_next(
     # ensure truncated dates do not have
     # time from 'now' included'
     if 'T' not in value.split(')')[0]:
-        cycle_point.hour_of_day = 0
-        cycle_point.minute_of_hour = 0
-        cycle_point.second_of_minute = 0
+        cycle_point._hour_of_day = 0
+        cycle_point._minute_of_hour = 0
+        cycle_point._second_of_minute = 0
+        # NOTE: Strictly speaking we shouldn't forcefully mutate TimePoints
+        # in this way as they're meant to be immutable since
+        # https://github.com/metomi/isodatetime/pull/165, however it
+        # should be ok as long as we don't call any of my_cp's methods
+        # (specifically, the ones that get cached) until after we've
+        # finished mutating it. I think.
+
     # ensure month and day from 'now' are not included
     # where they did not appear in the truncated datetime
     # NOTE: this may break when the order of tick over
     # for time point is reversed!!!
     # https://github.com/metomi/isodatetime/pull/101
     # case 1 - year only
     if re.search(r"\(-\d{2}[);T]", value):
-        cycle_point.month_of_year = 1
-        cycle_point.day_of_month = 1
+        cycle_point._month_of_year = 1
+        cycle_point._day_of_month = 1
     # case 2 - month only or year and month
     elif re.search(r"\(-(-\d{2}|\d{4})[;T)]", value):
-        cycle_point.day_of_month = 1
+        cycle_point._day_of_month = 1
```

Comment:
Should we create a new cycle point instead?

Reply:
I think this is the cleanest option for now. In future we could make a method on `isodatetime.TimePoint` to copy and replace kwargs, .e.g
```python
a = TimePoint(year=2022)
b = a.copy(month_of_year=4)
```

Reply:
We should be able to implement a clone method at this end for now. I think this will do?:

```python
from metomi.isodatetime.data import TimePoint


def clone_time_point(time_point):
    return TimePoint(
        **{
            attr: getattr(time_point, attr)
            for attr in TimePoint.DATA_ATTRIBUTES
            if attr != 'time_zone'
        },
        **{
            'time_zone_hour': time_point.time_zone.hours,
            'time_zone_minute': time_point.time_zone.minutes,
        },
    )
```

Reply:
There is already a `._copy()` method on `TimePoint` that does that. It just doesn't support changing an property. Adding the ability to change a property on copy would make it worthwhile to make it a public method IMHO. Otherwise, at the moment there is no gain to be had from making a copy here, as the `TimePoint` is a local variable in the function and not a dictionary key, so it is safe to mutate it in this circumstance

Reply:
The caveats you've outlined on hashing and cashing seem a bit scary to me and we *are* performing calculations on the timepoint which *could* potentially hit cached interfaces or force the hash to be calculated somehow?:

```
my_diff = [abs(my_time - now) for my_time in timepoints]
```

Seems like it should be easy enough to side-step this problem?

```python
def clone_time_point(time_point, **kwargs):
    return TimePoint(
        **{
            attr: getattr(time_point, attr)
            for attr in TimePoint.DATA_ATTRIBUTES
            if attr != 'time_zone'
        },
        **{
            'time_zone_hour': time_point.time_zone.hours,
            'time_zone_minute': time_point.time_zone.minutes,
        },
        **kwargs,
    )

clone_time_point(time_point, month_of_year=1, day_of_month=1)
```

Yes, we should add that to isodatetime one day but that can come later.

Reply:
We don't actually cache any methods yet - https://github.com/metomi/isodatetime/issues/176. So in isodatetime 3.1 we can add both caching and the `.copy()` method that supports changing attributes.

Reply:
I'm a little afraid of passing the attributes through `__init__()` again, I'm not 100% sure they will be unaffected. The `._copy()` method currently creates a new `TimePoint` bypassing `__init__()` and setting the attributes directly https://github.com/metomi/isodatetime/blob/731008b51346e75f3ca553c5b924586df6f69a34/metomi/isodatetime/data.py#L1621

Reply:
Not yet, and I guess we should create an issue for updating to the copy method when it is implemented

Reply:
Yep, please crack open an issue for that. I see you've fixed up the comment about caching, I think I'm happy this is ok.